pysr/regressor_sequence.py

```diff
@@ -0,0 +1,299 @@
+import warnings
+from typing import List, Optional, Union
+
+import numpy as np
+from sklearn.base import BaseEstimator
+
+from .sr import PySRRegressor
+from .utils import ArrayLike
+
+
+def _check_assertions(
+    X,
+    recursive_history_length=None,
+    weights=None,
+    variable_names=None,
+    X_units=None,
+):
+    if recursive_history_length is not None and recursive_history_length <= 0:
+        raise ValueError(
+            "The `recursive_history_length` parameter must be greater than 0 (otherwise it's not recursion)."
+        )
+    if len(X.shape) > 2:
+        raise ValueError(
+            "Recursive symbolic regression only supports up to 2D data; please flatten your data first"
+        )
+    if len(X) <= recursive_history_length + 1:
+        raise ValueError(
+            f"Recursive symbolic regression with a history length of {recursive_history_length} requires at least {recursive_history_length + 2} datapoints."
+        )
+    if isinstance(weights, np.ndarray) and len(weights) != len(X):
+        raise ValueError("The length of `weights` must have shape (n_times,).")
+    if isinstance(variable_names, list) and len(variable_names) != X.shape[1]:
+        raise ValueError(
+            "The length of `variable_names` must be equal to the number of features in `X`."
+        )
+    if isinstance(X_units, list) and len(X_units) != X.shape[1]:
+        raise ValueError(
+            "The length of `X_units` must be equal to the number of features in `X`."
+        )
+
+
+class PySRSequenceRegressor(BaseEstimator):
+    """
+    High performance symbolic regression for recurrent sequences.
+    Based off of the `PySRRegressor` class, but with a preprocessing step for recurrence relations.
+
+    Parameters
+    ----------
+    recursive_history_length : int
+        The number of previous time points to use as input features.
+        For example, if `recursive_history_length=2`, then the input features
+        will be `[X[0], X[1]]` and the output will be `X[2]`.
+        This continues on for all X: [X[n-1], X[n-2]] to predict X[n].
+        Must be greater than 0.
+    Other parameters and attributes are inherited from `PySRRegressor`.
+    """
+
+    def __init__(
+        self,
+        *,
+        recursive_history_length: int = 0,
+        **kwargs,
+    ):
+        super().__init__()
+        self._regressor = PySRRegressor(**kwargs)
+        self.recursive_history_length = recursive_history_length
+
+    def _construct_variable_names(
+        self, n_features: int, variable_names: Optional[List[str]]
+    ):
+        if not isinstance(variable_names, list):
+            if n_features == 1:
+                return [f"xt_{i}" for i in range(self.recursive_history_length, 0, -1)]
+            else:
+                return [
+                    f"x{i}t_{j}"
```

Comment:
Or even `x1_t1, x1_t2, ...`. This might mesh well with when the user provides custom features, like `a, b, c` would become `a_t1, a_t2, ..., b_t1, b_t2, ..., c_t1, c_t2, ...`.

Reply:
oh yeah the original idea was to use the underscore as a minus, but looking back its not really clear.